Keep compatibility since Emacs has removed typo.

nailuoGG opened the pull request:
the missing `chinse-two-byte` will break the main process, since emacs 31 remove the typo,  

https://github.com/emacs-mirror/emacs/commit/4437d730a5fff6dca61f99a146c99aefd0f67577

 fix #45

brsvh commented on pangu-spacing.el:
IMO, define an alias for `chinse-two-byte` to keep compatibility would be better.

nailuoGG replied:
You're right ,I think.

Emacs's compatibility advantage is a significant benefit of the ecosystem. Therefore, keeping compatibility is a more reasonable solution when dealing with outdated code

nailuoGG replied:
Emacs 24.3, [released in 2013](https://lists.gnu.org/archive/html/info-gnu-emacs/2013-03/msg00001.html), has already been around for 10 years, and given the modern features introduced after Emacs 28, it's reasonable to expect correct spellings.

brsvh replied:
But `rx--categories` is a constant variable, create an alias is hard.

I don't know how to hack `rx`, `rx--translate-form`, `rx--translate-category` for use the modified `rx--categories` to provide `chinse-two-byte`.

May be directly drop typo is better.

brsvh replied:
> Emacs 24.3, [released in 2013](https://lists.gnu.org/archive/html/info-gnu-emacs/2013-03/msg00001.html), has already been around for 10 years, and given the modern features introduced after Emacs 28, it's reasonable to expect correct spellings.

I mean keep compatibility for `pangu-spacing`, not the GNU Emacs.  As I know, most of user used this package did not use the latest Emacs.

nailuoGG replied:
> But `rx--categories` is a constant variable, create an alias is hard.
> 
> I don't know how to hack `rx`, `rx--translate-form`, `rx--translate-category` for use the modified `rx--categories` to provide `chinse-two-byte`.
> 
> May be directly drop typo is better.

I've been trying to use eval to achieve this, and the code works in Emacs 31 but hasn't been tested in Emacs 29 and Emacs 24.

```
(defun pangu-spacing-chinese-two-byte-category ()
  "Return the correct category name for Chinese two-byte characters based on Emacs version."
  (if (version< emacs-version "24.3")
      "chinse-two-byte"
    "chinese-two-byte"))

(defvar pangu-spacing-include-regexp
  (eval
   `(rx (or (and (or (group-n 3 (any "。，！？；：「」（）、"))
                     (group-n 1 (or (category ,(intern (pangu-spacing-chinese-two-byte-category)))
                                    (category japanese-hiragana-two-byte)
                                    (category japanese-katakana-two-byte))))
                 (group-n 2 (in "a-zA-Z0-9")))
            (and (group-n 1 (in "a-zA-Z0-9"))
                 (or (group-n 3 (any "。，！？；：「」（）、"))
                     (group-n 2 (or (category ,(intern (pangu-spacing-chinese-two-byte-category)))
                                    (category japanese-hiragana-two-byte)
                                    (category japanese-katakana-two-byte))))))))
  "Regexp to find Chinese character before English character.")


```

brsvh replied:
```
(if (version< emacs-version "24.3")
      "chinse-two-byte"
    "chinese-two-byte")
```

As your referenced commit, may be we should check `31.1` or `31.0.50` and later?

```
(if (not (version<= emacs-version "31.0"))
  ...
```

brsvh replied:
I don't have machine installed Emacs 24.3.

I have checked Emacs 28.2/29.4 in Nixpkgs, it seems work and no side effects.

nailuoGG replied:
https://github.com/coldnew/pangu-spacing/blob/2303013e5cd7852136f1429162fea0e1c8cb0221/pangu-spacing.el#L162-L169

Emacs versions 21.1 to 24.3 have incorrect spellings, later versions ( above 24.3 ) may include both correct and incorrect spellings (not sure).

Latest versions keep the correct spellings and remove the incorrect ones.

nailuoGG replied:
It doesn't seem like a good idea to hardcode the version number, but I couldn't find a way to detect the current available `rx--categories`, maybe it's just something I've missed?

brsvh replied:
> but I couldn't find a way to detect the current available rx--categories

Have you try `assq`?

```
(assq 'chinese-two-byte rx--categories)
```

nailuoGG replied:
I will combine the above content into a single commit. 😄

nailuoGG commented:
It's time to review the commit that sums up the updates we've discussed so far. 😄

nailuoGG committed:
Keep compatibility since Emacs has removed typo.

the missing `chinse-two-byte` may broken the package

https://github.com/emacs-mirror/emacs/commit/4437d730a5fff6dca61f99a146c99aefd0f67577

brsvh commented on pangu-spacing.el:
```diff
@@ -167,19 +167,25 @@ When you set t here, the space will be insert when you save file."
+;;
+(defun pangu-spacing-chinese-two-byte-category ()
+  "Return the correct category name for Chinese two-byte characters."
+  (if (assq 'chinese-two-byte rx--categories)
```
You also should require `rx` when byte compile, to inhibit error `Error: Symbol's value as variable is void: rx--categories`.

```
(eval-when-compile (require 'rx))
```

nailuoGG replied:
Fixed😄

nailuoGG committed:
fix missing package
